### PR TITLE
Filtering overhaul

### PR DESCRIPTION
**Pull request by dralley**
Fixes 

* https://pulp.plan.io/issues/3082
* https://pulp.plan.io/issues/3557
* https://pulp.plan.io/issues/3558

### CODE REVIEW

File: pulpcore/pulpcore/app/viewsets/base.py [hunk 1]
```diff
@@ -14,6 +14,9 @@
 from rest_framework.generics import get_object_or_404
 from rest_framework.schemas import AutoSchema
 
+NAME_FILTER_OPTIONS = ['exact', 'in']
+DATETIME_FILTER_OPTIONS = ['lt', 'lte', 'gt', 'gte', 'range']
```
Comment (dralley) [inline pulpcore/pulpcore/app/viewsets/base.py]: Other options:  "day", "week", "month", "day__range", "minute__range", "week__range"...

I had to restrain myself from going crazy after I figured out that that shit fuckin' works!

We can totally do a query like ```?created__hour__range=14,17``` if we want to enable that.

Comment (asmacdo) [reply pulpcore/pulpcore/app/viewsets/base.py]: Could you add some comments with example usage?

File: pulpcore/pulpcore/app/viewsets/base.py [hunk 2]
```diff
@@ -14,6 +14,9 @@
 from rest_framework.generics import get_object_or_404
 from rest_framework.schemas import AutoSchema
 
+NAME_FILTER_OPTIONS = ['exact', 'in']
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/base.py]: Wow, nice catch, "in" is indeed valid.
https://docs.djangoproject.com/en/dev/ref/models/querysets/#in

This class threw me off:
http://django-filter.readthedocs.io/en/master/ref/filters.html#baseinfilter

This DRYs things up quite a bit.

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 3]
```diff
@@ -186,19 +146,17 @@ def filter(self, qs, value):
 
 
 class RepositoryVersionFilter(filterset.FilterSet):
-
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/repository.py]: Some example uses would be nice.

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 4]
```diff
@@ -287,6 +245,27 @@ def create(self, request, repository_pk):
         return OperationPostponedResponse([result], request)
 
 
+class RemoteFilter(filterset.FilterSet):
+    """
+    Plugin remote filter would need:
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/repository.py]: s/filter would need/filters should/

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 5]
```diff
@@ -287,6 +245,27 @@ def create(self, request, repository_pk):
...
+     - to inherit from this class
+     - to add any specific filters if needed
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/repository.py]: - add type-specific filters 

If specific filters are not needed, they needn't subclass.

File: pulpcore/pulpcore/app/viewsets/repository.py (hunk 4 above)
Comment (asmacdo) [reply pulpcore/pulpcore/app/viewsets/repository.py]: also remove "to" from lines below.

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 6]
```diff
@@ -287,6 +245,27 @@ def create(self, request, repository_pk):
...
+     - to define its own `Meta` class which needs:
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/repository.py]: s/its own/a/
s/which needs/which should/
remove "to" from lines below.

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 7]
```diff
@@ -299,6 +278,27 @@ class RemoteViewSet(NamedModelViewSet,
     filter_class = RemoteFilter
 
 
+class PublisherFilter(filterset.FilterSet):
+    """
+    Plugin publisher filter would need:
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/repository.py]: same as above

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 8]
```diff
@@ -311,6 +311,18 @@ class PublisherViewSet(NamedModelViewSet,
     filter_class = PublisherFilter
 
 
+class ExporterFilter(filterset.FilterSet):
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/repository.py]: add similar docstring as above.

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 9]
```diff
@@ -332,6 +344,18 @@ class PublicationViewSet(NamedModelViewSet,
...
 
+class DistributionFilter(filterset.FilterSet):
+    name = filters.CharFilter()
+    base_path = filters.CharFilter()
+
+    class Meta:
+        model = Distribution
+        fields = {
+            'name': NAME_FILTER_OPTIONS,
+            'base_path': ['exact', 'in']
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/repository.py]: +contains?

File: pulpcore/pulpcore/app/viewsets/task.py [hunk 10]
```diff
@@ -5,26 +5,26 @@
...
+    finished_at = filters.IsoDateTimeFilter(name='finished_at')
 
     class Meta:
         model = Task
-        fields = ('state', 'state_in_list', 'worker',
-                  'started_after', 'started_before',
-                  'finished_after', 'finished_before')
+        fields = {
+            'state': NAME_FILTER_OPTIONS,
+            'worker': ['exact'],
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/task.py]: +in. Users will want to filter by all final and all non-final states.

File: pulpcore/pulpcore/app/viewsets/base.py (hunk 2 above)
Comment (dralley) [reply pulpcore/pulpcore/app/viewsets/base.py]: It surprised me to.  It works *automagically*

**Comment by asmacdo** [comment]
Other than requests for additional comments, requested changes are optional.

File: pulpcore/pulpcore/app/viewsets/repository.py (hunk 9 above)
Comment (dralley) [reply pulpcore/pulpcore/app/viewsets/repository.py]: +1

File: pulpcore/pulpcore/app/viewsets/task.py (hunk 10 above)
Comment (dralley) [reply pulpcore/pulpcore/app/viewsets/task.py]: Is this w/r/t line 23 or line 24?

NAME_FILTER_OPTIONS does include "in"

Comment (asmacdo) [reply pulpcore/pulpcore/app/viewsets/task.py]: I guess my eyes cross multiplied. Dunno what I was thinking lol.

File: pulpcore/pulpcore/app/viewsets/base.py (hunk 1 above)
Comment (dralley) [reply pulpcore/pulpcore/app/viewsets/base.py]: +1

File: pulpcore/pulpcore/app/viewsets/base.py (hunk 2 above)
Comment (asmacdo) [reply pulpcore/pulpcore/app/viewsets/base.py]: Aha! I did my initial work on filtering before this issue was filed, and did exactly what the reporter did.

File: pulpcore/pulpcore/app/viewsets/base.py [hunk 11]
```diff
@@ -14,6 +14,16 @@
...
 from rest_framework.schemas import AutoSchema
 
+# These should be used to prevent duplication and keep things consistent
+NAME_FILTER_OPTIONS = ['exact', 'in']
+# e.g.
+# /?name=foo
+# /?name__in=foo,bar
+DATETIME_FILTER_OPTIONS = ['lt', 'lte', 'gt', 'gte', 'range']
+# e.g.
+# /?created__gte=2018-04-12T19:45:52
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/base.py]: When I tried these filters forever ago, I also found them hard to use. This comment is helpful, but it is still unclear what the rules are.

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 12]
```diff
@@ -299,6 +283,26 @@ class RemoteViewSet(NamedModelViewSet,
     filter_class = RemoteFilter
 
 
+class PublisherFilter(filterset.FilterSet):
+    """
+    Plugin publisher filter should:
+     - inherit from this class
+     - add any specific filters if needed
+     - define a `Meta` class which needs:
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/repository.py]: s/needs/should/

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 13]
```diff
@@ -311,6 +315,26 @@ class PublisherViewSet(NamedModelViewSet,
     filter_class = PublisherFilter
 
 
+class ExporterFilter(filterset.FilterSet):
+    """
+    Plugin exporter filter should:
+     - inherit from this class
+     - add any specific filters if needed
+     - define a `Meta` class which needs:
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/repository.py]: same

File: pulpcore/pulpcore/app/viewsets/task.py [hunk 14]
```diff
@@ -5,26 +5,26 @@
...
+    started_at = filters.IsoDateTimeFilter(name='started_at')
+    finished_at = filters.IsoDateTimeFilter(name='finished_at')
 
     class Meta:
         model = Task
-        fields = ('state', 'state_in_list', 'worker',
-                  'started_after', 'started_before',
-                  'finished_after', 'finished_before')
+        fields = {
+            'state': ['exact', 'in'],
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/task.py]: This is more explicit for morons like me :)

**Pull request merged**